File: Machine Learning A-Z Template Folder/Part 2 - Regression/Section 8 - Decision Tree Regression/Exercises/decision_tree_regression.py
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
from sklearn.tree import DecisionTreeRegressor
from sklearn.preprocessing import StandardScaler

#-------------------------------------------------------------------------------
#                            Loading in Data
#-------------------------------------------------------------------------------

dataset = pd.read_csv('Position_Salaries.csv')
print('\n-------------------------------------------------------------------\n')
print("Head of dataset:\n",dataset.head())
print('\n-------------------------------------------------------------------\n')
x = dataset.iloc[:, 1:2].values
y = dataset.iloc[:, 2:3].values
print("x:\n", x)
print('\n-------------------------------------------------------------------\n')
print("y:\n",y)
print('\n-------------------------------------------------------------------\n')

#-------------------------------------------------------------------------------
#                  Creating Decision Tree Regressor
#-------------------------------------------------------------------------------

regressor = DecisionTreeRegressor(random_state = 0)
regressor.fit(x, y)
pred_65 = regressor.predict(6.5)
pred_62 = regressor.predict(6.2)
print(pred_65, pred_62)
print('\n-------------------------------------------------------------------\n')

#-------------------------------------------------------------------------------
#                  Visualizing Decision Tree Regression
#-------------------------------------------------------------------------------
x_grid = np.arange(min(x), max(x), 0.01)
x_grid = x_grid.reshape((len(x_grid), 1))

plt.scatter(x, y, color = 'r')
plt.plot(x_grid, regressor.predict(x_grid), c = 'blue')
plt.title("Truth or Bluff (Decision Tree Regression)")
plt.xlabel('Position level')
plt.ylabel('Salary')
plt.show()

File: Machine Learning A-Z Template Folder/Part 2 - Regression/Section 8 - Decision Tree Regression/Exercises/decision_tree_regression.R
# Decision Tree Regression

library(e1071)
library(ggplot2)
library(rpart)

#-------------------------------------------------------------------------------
#                            Loading in Data
#-------------------------------------------------------------------------------

dataset = read.csv('Position_Salaries.csv')
dataset = dataset[2:3]
print('-------------------------------------------------------------------')
print(dataset)
print('-------------------------------------------------------------------')
summary(dataset)
print('-------------------------------------------------------------------')

#-------------------------------------------------------------------------------
#                 Fitting Decision Tree Regression to the dataset
#-------------------------------------------------------------------------------

regressor = rpart(formula = Salary ~ .,
                data = dataset,
                control = rpart.control(minsplit = 3))
summary(regressor)

#-------------------------------------------------------------------------------
#             Predicting values for Decision Tree Regression
#-------------------------------------------------------------------------------

y_pred = predict(regressor, data.frame(Level = 6.5))
print("=======================================")
print(y_pred)
print("=======================================")

#-------------------------------------------------------------------------------
#             Visualizing Decision Tree Regression model to the dataset
#-------------------------------------------------------------------------------

png('decision_tree_regression.png')
x_grid = seq(min(dataset$Level), max(dataset$Level), 0.01)
decision_tree_plot = ggplot() +
  geom_point(aes(x = dataset$Level, y = dataset$Salary), colour = 'red') +
  geom_line(aes(x = x_grid, y = predict(regressor, newdata = data.frame(Level=x_grid))), colour = 'blue') +
  ggtitle('Salary prediction (Decision Tree Regression)') +
  xlab('Level') +
  ylab('Salary')
plot(decision_tree_plot)
dev.off()
